nemo_aligner/algorithms/spin.py

```diff
@@ -113,6 +115,14 @@ def __init__(
         self.num_steps_per_epoch = compute_num_steps_per_epoch(
             self.train_dataloader.batch_sampler, self.cfg.get("limit_train_batches", 1.0)
         )
+        
+        if isinstance(self.cfg.get("limit_train_batches", 1.0), int):
+            self.train_dataloader.batch_sampler.total_samples = min(
+                self.train_dataloader.batch_sampler.total_samples,
+                self.cfg.limit_train_batches * self.train_dataloader.batch_sampler.global_batch_size,
+            )
+            if hasattr(self.train_dataloader.batch_sampler, "last_batch_size"):
+                self.train_dataloader.batch_sampler.last_batch_size = 0
```

Comment:
You're correct, this is a really ugly hack because the Sampler class will have incorrect `total_samples` if you limit the training batches, and there will be a really ugly bug where your epochs don't align with the Sampler class, leading to a crash. The philosophically correct solution is to modify the Sampler class where the `total_samples` variable lives, which is what you recommended that I do, hence I created the Nemo PR.

Reply:
> The philosophically correct solution is to modify the Sampler class where the total_samples variable lives, which is what you recommended that I do, hence I created the Nemo PR

I agree we should modify the `total_samples` from the Sampler, but it seems to me we don't need to modify the Sampler class in NeMo -- we can instead provide the "corrected" value of `total_samples` (i.e., after accounting for `limit_train_batches`) to the Sampler, right? (reason I have a slight preference for this is it avoids having to touch NeMo at all).

Reply:
I see what you mean, yes, I think we can just move this logic to build_dataloader and kill off the Nemo PR

Reply:
Sounds good!

Reply:
I have moved this logic to `build_dataloader`